src/metabase/query_processor/store.clj

```diff
@@ -157,21 +157,19 @@
   [database-id-or-metadata-providerable :- ::database-id-or-metadata-providerable
    thunk                                :- [:=> [:cat] :any]]
   (cond
-    (or (not (initialized?))
-        *DANGER-allow-replacing-metadata-provider*)
-    (binding [*store*                                        (atom {})
-              *DANGER-allow-replacing-metadata-provider* false]
+    (not (initialized?))
+    (binding [*store* (atom {})]
       (do-with-metadata-provider database-id-or-metadata-providerable thunk))
 
-    ;; existing provider
-    (miscellaneous-value [::metadata-provider])
+    (or *DANGER-allow-replacing-metadata-provider*
+        (not (miscellaneous-value [::metadata-provider])))
     (do
-      (validate-existing-provider database-id-or-metadata-providerable)
+      (set-metadata-provider! database-id-or-metadata-providerable)
```

Comment:
Should we re-add the `binding [*DANGER-allow-replacing-metadata-provider* false]` from line 163 on the old version. To make sure the metadata provider is can only be rebound once unless someone explicitly opts into rebinding it again.

Reply:
Good call!